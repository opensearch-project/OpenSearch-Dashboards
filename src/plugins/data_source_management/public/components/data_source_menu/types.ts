/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  MountPoint,
  NotificationsStart,
  SavedObjectsClientContract,
  SavedObject,
  IUiSettingsClient,
} from '../../../../../core/public';
import { DataSourceAttributes } from '../../types';

export interface DataSourceOption {
  id: string;
  label?: string;
}

export interface DataSourceGroupLabelOption extends DataSourceOption {
  label: string;
  isGroupLabel: true;
}

export interface DataSourceBaseConfig {
  fullWidth: boolean;
  disabled?: boolean;
}

export interface DataSourceMenuProps<T = any> {
  componentType: DataSourceComponentType;
  componentConfig: T;
  hideLocalCluster?: boolean;
  uiSettings?: IUiSettingsClient;
  setMenuMountPoint?: (menuMount: MountPoint | undefined) => void;
}

export const DataSourceComponentType = {
  DataSourceSelectable: 'DataSourceSelectable',
  DataSourceView: 'DataSourceView',
  DataSourceAggregatedView: 'DataSourceAggregatedView',
  DataSourceMultiSelectable: 'DataSourceMultiSelectable',
} as const;

export type DataSourceComponentType = typeof DataSourceComponentType[keyof typeof DataSourceComponentType];

export interface DataSourceViewConfig extends DataSourceBaseConfig {
  activeOption: DataSourceOption[];
  savedObjects?: SavedObjectsClientContract;
  notifications?: NotificationsStart;
}

export interface DataSourceAggregatedViewConfig extends DataSourceBaseConfig {
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
  activeDataSourceIds?: string[];
  displayAllCompatibleDataSources?: boolean;
  dataSourceFilter?: (dataSource: SavedObject<DataSourceAttributes>) => boolean;
}

export interface DataSourceSelectableConfig extends DataSourceBaseConfig {
  onSelectedDataSources: (dataSources: DataSourceOption[]) => void;
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
  activeOption?: DataSourceOption[];
  dataSourceFilter?: (dataSource: SavedObject<DataSourceAttributes>) => boolean;
}

export interface DataSourceMultiSelectableConfig extends DataSourceBaseConfig {
  onSelectedDataSources: (dataSources: DataSourceOption[]) => void;
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
}
