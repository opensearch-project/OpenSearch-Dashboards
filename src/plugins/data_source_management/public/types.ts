/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ChromeStart,
  ApplicationStart,
  IUiSettingsClient,
  OverlayStart,
  SavedObjectsStart,
  NotificationsStart,
  DocLinksStart,
  HttpSetup,
} from 'src/core/public';
import { ManagementAppMountParams } from 'src/plugins/management/public';
import { SavedObjectAttributes } from 'src/core/types';
import { i18n } from '@osd/i18n';
import { SigV4ServiceName } from '../../data_source/common/data_sources';
import { OpenSearchDashboardsReactContextValue } from '../../opensearch_dashboards_react/public';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataSourceManagementPluginStart {}

export interface DataSourceManagementContext {
  chrome: ChromeStart;
  application: ApplicationStart;
  savedObjects: SavedObjectsStart;
  uiSettings: IUiSettingsClient;
  notifications: NotificationsStart;
  overlays: OverlayStart;
  http: HttpSetup;
  docLinks: DocLinksStart;
  setBreadcrumbs: ManagementAppMountParams['setBreadcrumbs'];
}

export interface DataSourceTableItem {
  id: string;
  title: string;
  description: string;
  sort: string;
}

export interface ToastMessageItem {
  id: string;
  defaultMessage: string;
  success?: boolean;
}

export type DataSourceManagementContextValue = OpenSearchDashboardsReactContextValue<
  DataSourceManagementContext
>;

/* Datasource types */
export enum AuthType {
  NoAuth = 'no_auth',
  UsernamePasswordType = 'username_password',
  SigV4 = 'sigv4',
}

export const credentialSourceOptions = [
  {
    value: AuthType.NoAuth,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.NoAuthentication', {
      defaultMessage: 'No authentication',
    }),
  },
  {
    value: AuthType.UsernamePasswordType,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.UsernamePassword', {
      defaultMessage: 'Username & Password',
    }),
  },
  {
    value: AuthType.SigV4,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.AwsSigV4', {
      defaultMessage: 'AWS SigV4',
    }),
  },
];

export const sigV4ServiceOptions = [
  {
    value: SigV4ServiceName.OpenSearch,
    text: i18n.translate('dataSourceManagement.SigV4ServiceOptions.OpenSearch', {
      defaultMessage: 'Amazon OpenSearch Service',
    }),
  },
  {
    value: SigV4ServiceName.OpenSearchServerless,
    text: i18n.translate('dataSourceManagement.SigV4ServiceOptions.OpenSearchServerless', {
      defaultMessage: 'Amazon OpenSearch Serverless',
    }),
  },
];

export interface DataSourceAttributes extends SavedObjectAttributes {
  title: string;
  description?: string;
  endpoint?: string;
  auth: {
    type: AuthType;
    credentials: UsernamePasswordTypedContent | SigV4Content | undefined;
  };
}

export interface UsernamePasswordTypedContent extends SavedObjectAttributes {
  username: string;
  password?: string;
}

export interface SigV4Content extends SavedObjectAttributes {
  accessKey: string;
  secretKey: string;
  region: string;
  service?: SigV4ServiceName;
}
