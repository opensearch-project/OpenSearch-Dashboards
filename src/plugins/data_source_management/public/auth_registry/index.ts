/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export {
  IAuthenticationMethodRegistry,
  AuthenticationMethod,
  AuthenticationMethodRegistry,
} from './authentication_methods_registry';
