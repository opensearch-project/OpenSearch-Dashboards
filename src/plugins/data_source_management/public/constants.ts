/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DirectQueryDatasourceType } from './types';

export const QUERY_RESTRICTED = 'query-restricted';
export const QUERY_ALL = 'query-all';

export const DatasourceTypeToDisplayName: { [key in DirectQueryDatasourceType]: string } = {
  PROMETHEUS: 'Prometheus',
  S3GLUE: 'Amazon S3',
};

export const PROMETHEUS_URL = 'Prometheus';
export const AMAZON_S3_URL = 'AmazonS3AWSGlue';
export const OPENSEARCH_URL = 'OpenSearch';

export const UrlToDatasourceType: { [key: string]: DirectQueryDatasourceType } = {
  [PROMETHEUS_URL]: 'PROMETHEUS',
  [AMAZON_S3_URL]: 'S3GLUE',
};
