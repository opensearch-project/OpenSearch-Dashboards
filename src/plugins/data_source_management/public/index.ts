/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataSourceManagementPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new DataSourceManagementPlugin();
}
export { DataSourceManagementPluginStart } from './types';
