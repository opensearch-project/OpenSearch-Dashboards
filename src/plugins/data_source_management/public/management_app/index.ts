/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { mountManagementSection } from './mount_management_section';
