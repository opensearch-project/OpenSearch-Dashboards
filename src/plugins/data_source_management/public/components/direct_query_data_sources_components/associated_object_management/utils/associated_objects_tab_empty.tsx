/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton, EuiEmptyPrompt, EuiText } from '@elastic/eui';
import React from 'react';
import { DataSourceManagementContext } from 'src/plugins/data_source_management/public/types';
import { LoadCacheType } from '../../../../../framework/types';
import { useOpenSearchDashboards } from '../../../../../../opensearch_dashboards_react/public';

interface AssociatedObjectsTabEmptyProps {
  cacheType: LoadCacheType;
}

export const AssociatedObjectsTabEmpty: React.FC<AssociatedObjectsTabEmptyProps> = (props) => {
  const { cacheType } = props;

  const { application } = useOpenSearchDashboards<DataSourceManagementContext>().services;

  const QueryWorkbenchButton = (
    <EuiButton
      iconSide="right"
      onClick={() => application.navigateToApp('opensearch-query-workbench')}
      iconType="popout"
    >
      Query Workbench
    </EuiButton>
  );

  let titleText;
  let bodyText;
  switch (cacheType) {
    case 'databases':
      titleText = 'You have no databases in your data source';
      bodyText = 'Add databases and tables to your data source or use Query Workbench';
      break;
    case 'tables':
      titleText = 'You have no associated objects';
      bodyText = 'Add tables to your data source or use Query Workbench';
      break;
    default:
      titleText = '';
      bodyText = '';
      break;
  }

  return (
    <EuiEmptyPrompt
      body={
        <>
          <EuiText>
            <h4>{titleText}</h4>
            <p>{bodyText}</p>
          </EuiText>
        </>
      }
      actions={QueryWorkbenchButton}
    />
  );
};
