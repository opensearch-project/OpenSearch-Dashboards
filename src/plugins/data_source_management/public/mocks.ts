/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { throwError } from 'rxjs';
import { SavedObjectsClientContract } from 'opensearch-dashboards/public';
import { AuthType } from './types';
import { coreMock } from '../../../core/public/mocks';

/* Mock Types */

export const docLinks = {
  links: {
    noDocumentation: {
      indexPatterns: {
        introduction: '',
      },
      scriptedFields: {},
    },
  },
};

const createDataSourceManagementContext = () => {
  const {
    chrome,
    application,
    savedObjects,
    uiSettings,
    notifications,
    overlays,
  } = coreMock.createStart();
  const { http } = coreMock.createSetup();

  return {
    chrome,
    application,
    savedObjects,
    uiSettings,
    notifications,
    overlays,
    http,
    docLinks,
    setBreadcrumbs: () => {},
  };
};

export const mockManagementPlugin = {
  createDataSourceManagementContext,
  docLinks,
};

/* Mock data responses - JSON*/
export const getDataSourcesResponse = {
  savedObjects: [
    {
      id: 'test',
      type: 'data-source',
      description: 'test datasource',
      title: 'test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'test2',
      type: 'data-source',
      description: 'test datasource2',
      title: 'test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'alpha-test',
      type: 'data-source',
      description: 'alpha test datasource',
      title: 'alpha-test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'beta-test',
      type: 'data-source',
      description: 'beta test datasource',
      title: 'beta-test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
  ],
};

export const existingDatasourceNamesList = [
  'test123',
  'testTest20',
  'TeSt',
  'duplicateTest',
  'dup20',
];

export const getMappedDataSources = [
  {
    id: 'test',
    description: 'test datasource',
    title: 'test',
    sort: 'test',
  },
  {
    id: 'test2',
    description: 'test datasource2',
    title: 'test',
    sort: 'test',
  },
  {
    id: 'alpha-test',
    description: 'alpha test datasource',
    title: 'alpha-test',
    sort: 'alpha-test',
  },
  {
    id: 'beta-test',
    description: 'beta test datasource',
    title: 'beta-test',
    sort: 'beta-test',
  },
];

export const mockDataSourceAttributesWithAuth = {
  id: 'test',
  title: 'create-test-ds',
  description: 'jest testing',
  endpoint: 'https://test.com',
  auth: {
    type: AuthType.UsernamePasswordType,
    credentials: {
      username: 'test123',
      password: 'test123',
    },
  },
};

export const mockDataSourceAttributesWithNoAuth = {
  id: 'test123',
  title: 'create-test-ds123',
  description: 'jest testing',
  endpoint: 'https://test.com',
  auth: {
    type: AuthType.NoAuth,
    credentials: undefined,
  },
};
export const getDataSourceByIdWithCredential = {
  attributes: {
    id: 'alpha-test',
    title: 'alpha-test',
    endpoint: 'https://test.com',
    description: 'alpha test data source',
    auth: {
      type: AuthType.UsernamePasswordType,
      credentials: {
        username: 'test123',
      },
    },
  },
};

export const getDataSourceByIdWithoutCredential = {
  attributes: {
    ...getDataSourceByIdWithCredential.attributes,
    auth: {
      type: AuthType.NoAuth,
      credentials: undefined,
    },
  },
  references: [],
};

export const mockResponseForSavedObjectsCalls = (
  savedObjectsClient: SavedObjectsClientContract,
  savedObjectsMethodName: 'get' | 'find' | 'create' | 'delete' | 'update',
  response: any
) => {
  (savedObjectsClient[savedObjectsMethodName] as jest.Mock).mockResolvedValue(response);
};

export const mockErrorResponseForSavedObjectsCalls = (
  savedObjectsClient: SavedObjectsClientContract,
  savedObjectsMethodName: 'get' | 'find' | 'create' | 'delete' | 'update'
) => {
  (savedObjectsClient[savedObjectsMethodName] as jest.Mock).mockRejectedValue(
    throwError(new Error('Error while fetching data sources'))
  );
};
