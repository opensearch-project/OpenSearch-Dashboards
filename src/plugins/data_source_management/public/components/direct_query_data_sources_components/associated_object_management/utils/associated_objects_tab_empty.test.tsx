/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { AssociatedObjectsTabEmpty } from './associated_objects_tab_empty';
import { EuiButton, EuiEmptyPrompt, EuiText } from '@elastic/eui';
import { useOpenSearchDashboards } from '../../../../../../opensearch_dashboards_react/public';
import { DataSourceManagementContext } from 'src/plugins/data_source_management/public/types';

// Mock useOpenSearchDashboards hook
jest.mock('../../../../../../opensearch_dashboards_react/public', () => ({
  useOpenSearchDashboards: jest.fn(),
}));

describe('AssociatedObjectsTabEmpty', () => {
  const mockNavigateToApp = jest.fn();
  const mockServices = {
    application: {
      navigateToApp: mockNavigateToApp,
    },
  };

  const mockUseOpenSearchDashboards = useOpenSearchDashboards as jest.Mock;
  mockUseOpenSearchDashboards.mockReturnValue({ services: mockServices });

  const defaultProps = {
    cacheType: 'databases' as 'databases' | 'tables',
  };

  const shallowComponent = (props = defaultProps) =>
    shallow(<AssociatedObjectsTabEmpty {...props} />);
  const mountComponent = (props = defaultProps) => mount(<AssociatedObjectsTabEmpty {...props} />);

  test('renders correctly', () => {
    const wrapper = shallowComponent();
    expect(wrapper).toMatchSnapshot();
  });

  test('displays the correct title and body text for databases', () => {
    const wrapper = mountComponent({ cacheType: 'databases' });
    expect(wrapper.find(EuiText).find('h4').text()).toEqual(
      'You have no databases in your data source'
    );
    expect(wrapper.find(EuiText).find('p').text()).toEqual(
      'Add databases and tables to your data source or use Query Workbench'
    );
  });

  test('displays the correct title and body text for tables', () => {
    const wrapper = mountComponent({ cacheType: 'tables' });
    expect(wrapper.find(EuiText).find('h4').text()).toEqual('You have no associated objects');
    expect(wrapper.find(EuiText).find('p').text()).toEqual(
      'Add tables to your data source or use Query Workbench'
    );
  });

  test('renders the Query Workbench button', () => {
    const wrapper = mountComponent();
    expect(wrapper.find(EuiButton).text()).toEqual('Query Workbench');
  });

  test('calls navigateToApp when Query Workbench button is clicked', () => {
    const wrapper = mountComponent();
    wrapper.find(EuiButton).simulate('click');
    expect(mockNavigateToApp).toHaveBeenCalledWith('opensearch-query-workbench');
  });
});
