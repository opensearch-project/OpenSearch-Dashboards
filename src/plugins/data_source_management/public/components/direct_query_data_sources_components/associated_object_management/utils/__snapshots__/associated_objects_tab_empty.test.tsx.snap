// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssociatedObjectsTabEmpty renders correctly 1`] = `
<EuiEmptyPrompt
  actions={
    <EuiButton
      iconSide="right"
      iconType="popout"
      onClick={[Function]}
    >
      Query Workbench
    </EuiButton>
  }
  body={
    <React.Fragment>
      <EuiText>
        <h4>
          You have no databases in your data source
        </h4>
        <p>
          Add databases and tables to your data source or use Query Workbench
        </p>
      </EuiText>
    </React.Fragment>
  }
/>
`;
